doc/condor_configure_pool.1
.TH CONDOR_CONFIGURE_POOL 1 "August 2011" condor-wallaby "User Commands"
.SH NAME
condor_configure_pool \- Create condor specific configurations for nodes in
wallaby
.SH SYNOPSIS
.B condor_configure_pool
[\fIOPTION\fR]... \fIACTION TARGET\fR
.br
.B condor_configure_pool
[\fIOPTION\fR]... \fIACTION TARGET ENTITY\fR...
.SH DESCRIPTION
.PP
condor_configure_pool is a condor specific interface into wallaby.  It
provides a guided interface to applying features and parameters to a node
or a group of nodes.  When editing a node or group the condor_configure_pool
tool uses a text editor to present the node or group metadata in YAML format.

.SH OPTIONS
.SS Options
.TP
\-b, --broker BROKER
The ip/hostname of the broker used by wallaby.
.TP
-o, --port PORT
The port of the broker used by wallaby.
.TP
-U, --user USER
The username used to authenticate with the AMQP broker.
.TP
-P, --password PASSWORD
The password used to authenticate with the AMQP broker.
.TP
-m, --auth-mechanism MECH
A comma separated list of authentication mechanisms to use when communicating
with the broker.  Supported mechanisms are: ANONYMOUS, PLAIN, GSSAPI.
.TP
--schedds
Prompt for scheduler information.
.TP
--qmfbroker
Prompt for QMF broker information.
.TP
-h, --help
Print help.
.TP
-v, --verbose
Print more information, if available.
.SS Action:
Only 1 action is allowed per command invocation.
.TP
-a, --add
Append to the group/node with lowest priority.
.TP
-d, --delete
Remove from the group/node.
.TP
-e, --edit
Edit the group/node.  Unlike other actions, this requires no entities.  If
entities are provided they will be ignored.
.TP
-i, --insert
Insert into the group/node with highest priority.
.TP
-l, --list
List detailed information about the node/group.  Unlike other actions, this
requires no entities.  If entities are provided they will be ignored.
.TP
--activate
Attempt to activate the current configuration in wallaby.
.TP
--load-snapshot NAME
Load snapshot then with the given name from wallaby.
.TP
--take-snapshot NAME
Create a snapshot with the given name from wallaby.
.TP
--remove-snapshot NAME
Remove the snapshot with the given name from wallaby.
.SS Special List Actions:
Unlike regular actions, more than 1 is allowed per command invocation.
However, these can not be mixed with regular actions in a single command
invocation.  These actions do not require a target.
.TP
--list-all-features
List all features in wallaby.
.TP
--list-all-groups
List all groups in wallaby.
.TP
--list-all-nodes
List all nodes in wallaby.
.TP
--list-all-params
List all params in wallaby.
.TP
--list-all-snapshots
List all snapshots in wallaby.
.SS Targets
.TP
--default-group
Perform actions on wallaby's default group.
.TP
--skel-group
Perform actions on wallaby's skeleton group.
.TP
-n, --node NAME
The name of a specific node that will have the configuration change
applied to it.
.TP
-g, --group NAME
The name of the group that will have the configuration change applied to it.

NOTE: "Internal Default Group" can be used as the group name to denote
wallaby's default group instead of using --default-group.
.SS Entities
Multiple entities can be acted upon in a single command invocation, and lists
of each entity are separated by commas.  If a name has a comma in it, then
the name will need to be surrounded by escaped quotes (\\").
.TP
-f, --features FEATURE[,FEATURE,...]
A comma separated list of feature names that will be acted upon.
.TP
-p, --params PARAM[,PARAM,...]
A comma separated list of parameter names that will be acted upon.  Values for
the parameters will be prompted for if they are being added.
.SH EXIT STATUS
.PP
The exit status is 0 if the action is performed successfully on all targets
and 1 otherwise.  Failure to find the name of an entity in wallaby is
considered a failure and will return 1.
.SH ENVIRONMENT
EDITOR
.RS
The editor that will be used to edit YAML representations of wallaby entities.
If not defined, vi will be used.
.SH NOTES
Use condor_configure_store to create entities in wallaby
.SH EXAMPLE
.TP
Adding features to a group:
.B condor_configure_pool -a -f FeatureA,FeatureB -g GroupA
.TP
Add features and parameters to the default group:
.B condor_configure_pool --default-group -a -f FeatureA,FeatureB -p ParamA,ParamB
.TP
Listing all of the nodes and features in wallaby:
.B condor_configure_pool --list-all-nodes --list-all-features
.TP
List node details over a secure AMQP broker:
.B condor_configure_pool -U bob -P notsosecure -l -n NodeA


condor_configure_pool.rb
#!/usr/bin/ruby
#   Copyright 2012 Red Hat, Inc.
#
#   Licensed under the Apache License, Version 2.0 (the "License");
#   you may not use this file except in compliance with the License.
#   You may obtain a copy of the License at
#
#       http://www.apache.org/licenses/LICENSE-2.0
#
#   Unless required by applicable law or agreed to in writing, software
#   distributed under the License is distributed on an "AS IS" BASIS,
#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#   See the License for the specific language governing permissions and
#   limitations under the License.
require 'rbconfig'
require 'optparse'
require 'mrg/grid/config/shell'

module Mrg
  module Grid
    module Config
      class CCP
        def store_types
          [:Feature, :Group, :Node, :Parameter, :Snapshot]
        end

        def entities
          [:Feature, :Parameter]
        end

        def actions
          [[:add, "Append to the group/node with lowest priority"],
           [:delete, "Remove from the group/node"],
           [:edit, "Edit the group/node"],
           [:insert, "Insert into the group/node with highest priority"],
           [:list, "List detailed information"],
           [:activate, "Attempt to activate the configuration in the store"]]
        end

        def no_target
          [:listall, :activate]
        end

        def no_entities
          [:list, :listall, :edit, :activate]
        end

        def parse_option_args(args)
          items = []
          quotes = false
          current = ""
          args.each_byte do |c|
            c = c.chr
            if (c == '"') || (c == "'"):
              if quotes == false
                quotes = true
                next
              else
                quotes = false
                next
              end
            elsif (c == ',') && (quotes == false)
              items.push(current)
              current = ""
              next
            end
            current << c
          end
          items.push(current)
          items
        end

        def main(args)
          ws_args = []
          verbose = false
          options = {}

          @op = OptionParser.new do |opts|
            opts.banner = "Usage: #{File.basename($0)} [options] <action> <target> [config entities]"

            opts.separator ""
            opts.separator "Options:"
            opts.on("-h", "--help", "shows this message") do
              puts @op
              exit(0)
            end

            opts.on("-b", "--broker HOSTNAME", "The qpid broker hostname used by the configuration store") do |h|
              ws_args << ["-H", h]
            end

            opts.on("-o", "--port NUM", Integer, "The qpid broker port used by the configuration store") do |p|
              ws_args << ["-p", p]
            end

            opts.on("-U", "--user NAME", "The username used to authenticate with the qpid broker") do |n|
              ws_args << ["-U", n]
            end

            opts.on("-P", "--password PASS", "The password used to authenticate with the qpid broker") do |p|
              ws_args << ["-P", p]
            end

            opts.on("-m", "--auth-mechanism PASS", %w{ANONYMOUS PLAIN GSSAPI}, "A comma separated list of authentication mechanisms (#{%w{ANONYMOUS PLAIN GSSAPI}.join(", ")}) for authenticating with the qpid broker") do |m|
              ws_args << ["-M", m]
            end

            opts.on("--schedds", "Prompt for scheduler information") do
              options[:schedds] = true
            end

            opts.on("--qmfbroker", "Prompt for QMF broker information") do
              options[:qmf] = true
            end

            opts.on("-v", "--verbose", "Print more information, if available") do
              verbose = true
            end

            opts.separator ""
            opts.separator "Target: (Only 1 allowed per invocation)"

            targets = [["", "default-group", "Perform actions on the Internal Default Group in the store", nil],
                       ["-n", "node", "The name of a specific machine that will have the configuration changes applied to it", "NAME"],
                       ["-g", "group", "The name of the group that will have the configuration changes applied to it", "NAME"],
                       ["", "skel-group", "Perform actions on the Skeleton Group in the store", nil]]
            targets.each do |short, long, desc, arg|
              s = "-#{short}" unless short.empty?
              opts.on(s, "--#{[long, arg].join(" ")}", desc) do |n|
                if options.has_key?(:target)
                  puts "Can only configure 1 node or group at a time"
                  exit(1)
                end
                t = long.split('-')
                options[:target] = t[0].capitalize.to_sym if t.length < 2
                options[:target] = t[1].capitalize.to_sym if t.length > 1
                options[:target_name] = n if n
                options[:target_name] = "+++#{long.split("-").first.upcase}" unless n
              end
            end

            opts.separator ""
            opts.separator "Action: (Only 1 allowed per invocation)"

            actions.each do |name, desc|
              short = "-#{name.to_s[0,1]}" unless name == :activate
              opts.on(short, "--#{name}", desc) do
                if options.has_key?(:action)
                  puts "Only 1 action may be specified"
                  exit(1)
                end
                options[:action] = name
                options[:action] = :remove if name == :delete
              end
            end

            ["load", "take", "remove"].each do |type|
              opts.on("--#{type}-snapshot NAME", "#{type.capitalize} snapshot with the given name") do |n|
                if options.has_key?(:action)
                  puts "Only 1 action may be specified"
                  exit(1)
                end
                options[:action] = :snapshot
                options[:snapshot] = {:type=>type.gsub(/t/, "m").to_sym, :name=>n}
              end
            end

            opts.separator ""
            opts.separator "List Action: (Cannot be combined with actions)"

            store_types.each do |type|
              if type == :Parameter
                str = "param"
              else
                str = type.to_s.downcase
              end
              opts.on("--list-all-#{str}s", "List all #{str}s in the store") do
                if options.has_key?(:action) && options[:action] != :listall
                  puts "Actions and List Actions cannot be combined"
                  exit(1)
                end
                options[str.capitalize.to_sym] = "" 
                options[:action] = :listall
              end
            end

            opts.separator ""
            opts.separator "Config Entities:"

            entities.each do |n|
              ln = n.to_s.slice(0,5).upcase
              ln.chop! if n == :Feature
              on = n.to_s.downcase + "s"
              on = ln.downcase + "s" if n == :Parameter
              opts.on("-#{n.to_s[0,1].downcase}", "--#{on} #{ln}[,#{ln},...]", "A comma separated list of #{n}s") do |list|
                if options.has_key?(n)
                  options[n] += "," + list
                else
                  options[n] = list
                end
              end
            end
          end

          begin
            @op.parse!(args)
          rescue OptionParser::InvalidOption
            puts @op
            exit(1)
          rescue OptionParser::InvalidArgument => ia
            puts ia
            puts @op
            exit(1)
          rescue OptionParser::AmbiguousOption => ao
            puts ao
            puts @op
            exit(1)
          rescue OptionParser::MissingArgument => ma
            puts ma
            puts @op
            exit(1)
          end

          unless options.has_key?(:action)
            puts "No action specified.  Exiting"
            puts @op
            exit(1)
          end

          opts = []
          opts.push("--schedds") if options.has_key?(:schedds)
          opts.push("--qmf") if options.has_key?(:qmf)
  
          # Generate the option string to pass to the shell command
          opts.push("#{options[:target]}=#{options[:target_name]}") if options.has_key?(:target)
          options.each_key do |type|
            next if not store_types.include?(type)
            parse_option_args(options[type]).each do |name|
              opts.push("#{type}=#{name}")
            end
          end
          cmd = "ccp-#{options[:action]}"
          if options[:action] == :snapshot
            cmd = "#{options[:snapshot][:type]}-#{options[:action]}"
            opts = [options[:snapshot][:name]]
          end
          if options[:action] == :activate
            cmd = "#{options[:action]}"
            opts = []
          end

          # Call wallaby shell ccp to perform the actions
          ENV['WALLABY_COMMAND_DIR'] = "#{Config::CONFIG["sitelibdir"]}/condor_wallaby_tools" unless ENV['WALLABY_COMMAND_DIR']
          Mrg::Grid::Config::Shell::install_commands()
          Mrg::Grid::Config::Shell::main(ws_args + [cmd] + opts.flatten)
        end
      end
    end
  end
end


Mrg::Grid::Config::CCP.new.main(ARGV)
